File: samples/ts/src/DoublyLinkedLists.ts
class DoublyLinkedLists {

    static runMain() {
        new DoublyLinkedLists().test();
    }

    public test() {
        // custom linkedlist
        const linkedList: DoublyLinkedList<string> = new DoublyLinkedList<string>();
        for (let i = 0; i < 7; i++) {
            linkedList.append(String(i));
        }

        let node: Node<string> | undefined = linkedList.getHead();
        while (node != null) {
            console.log(node.getValue());
            if (node.getValue() === "5") {
                node.insertNext(new Node<string>("50"));
                break;
            }
            node = node.getNext();
        }
    }
}

export class DoublyLinkedList<T> {

    private head: Node<T> | undefined;
    private tail: Node<T> | undefined;

    public getHead(): Node<T> | undefined {
        return this.head;
    }

    public getTail(): Node<T> | undefined {
        return this.tail;
    }

    public append(value: T) {
        if (this.head == null) {
            this.head = new Node<T>(value);
            this.tail = this.head;
        } else {
            this.tail?.insertNext(new Node<T>(value));
            this.tail = this.tail?.getNext();
        }
    }
}

export class Node<T> {

    private value: T;
    private next: Node<T> | undefined;
    private prev: Node<T> | undefined;

    public constructor(value: T) {
        this.value = value;
    }

    public getValue(): T {
        return this.value;
    }

    public getNext(): Node<T> | undefined {
        return this.next;
    }

    public getPrev(): Node<T> | undefined {
        return this.prev;
    }

    public insertNext(node: Node<T>) {
        const t: Node<T> | undefined = this.next;
        this.next = node;
        node.next = t;
        if (t != null)
            t.prev = node;
        node.prev = this;
    }

    public toString(): string {
        return String(this.value);
    }
}

DoublyLinkedLists.runMain();

File: samples/python/src/DoublyLinkedLists.py
from __future__ import annotations
from typing import TypeVar, Generic
from collections import deque

T = TypeVar("T")


class DoublyLinkedLists:

    @staticmethod
    def run_main():
        DoublyLinkedLists().start()

    def start(self):
        # custom linkedlist
        linkedList = DoublyLinkedLists.DoublyLinkedList[str]()
        for i in range(10):
            linkedList.append(str(i))

        node: DoublyLinkedLists.Node[str] = linkedList.get_head()
        while node != None:
            print(node.get_value())
            if node.get_value() == "5":
                node.insert_next(DoublyLinkedLists.Node[str]("50"))
                break
            node = node.get_next()
        print("done")

    class DoublyLinkedList(Generic[T]):

        def __init__(self):
            self.head: DoublyLinkedLists.Node[T] = None
            self.tail: DoublyLinkedLists.Node[T] = None

        def get_head(self) -> DoublyLinkedLists.Node[T]:
            return self.head

        def get_tail(self) -> DoublyLinkedLists.Node[T]:
            return self.tail

        def append(self, value: T):
            if self.head is None:
                self.head = DoublyLinkedLists.Node[T](value)
                self.tail = self.head
            else:
                self.tail.insert_next(DoublyLinkedLists.Node[T](value))
                self.tail = self.tail.next

    class Node(Generic[T]):

        def __init__(self, value: T):
            self.value: T = value
            self.next: DoublyLinkedLists.Node[T] = None
            self.prev: DoublyLinkedLists.Node[T] = None

        def get_value(self) -> T:
            return self.value

        def get_next(self) -> DoublyLinkedLists.Node[T]:
            return self.next

        def get_prev(self) -> DoublyLinkedLists.Node[T]:
            return self.prev

        def insert_next(self, node: DoublyLinkedLists.Node[T]):
            t: DoublyLinkedLists.Node[T] = self.next
            self.next = node
            node.next = t
            if t is not None:
                t.prev = node
            node.prev = self

        def __str__(self) -> str:
            return str(self.value)

DoublyLinkedLists.run_main()
